cibyl/features/__init__.py

```diff
@@ -0,0 +1,156 @@
+                # classes defined there
+                # https://docs.python.org/3/library/importlib.html#importing-a-source-file-directly
+                spec = spec_from_file_location(module_name+"_features",
+                                               f"{location}/{module_path}")
+                module = module_from_spec(spec)
+                spec.loader.exec_module(module)
+                features = getmembers(module, predicate=is_feature_class)
+                for feature_name, feature in features:
+                    self.all_features[feature_name.lower()] = feature
+                    self.features_by_category[module_name].append(feature_name)
+
+    def get_string_all_features(self):
+        """Get a string representation listing all available features to use in
+        an exception message."""
+        msg = ""
+        for category, features_names in self.features_by_category.items():
+            msg += f"\n{Colors.blue(category.title())}:"
+            for feature_name in features_names:
+                feature = self.all_features[feature_name.lower()]
+                docstring = getattr(feature, "__doc__", "")
+                msg += f"\n{Colors.red(' * ')}{Colors.blue(feature_name)}"
+                if docstring:
+                    msg += f"{Colors.red(' - '+feature.__doc__)}"
+        return msg
+
+    def get_feature(self, name_feature):
+        """Get the function associated with the given feature name
+
+        :param feature_name: Name of the feature
+        :type feature_name: str
+
+        :returns: class that implements the given feature
+        :rtype: class
+        :raises: InvalidArgument
+        """
+        try:
+            feature_class = self.all_features[name_feature.lower()]
+        except KeyError as err:
+            msg = f"No feature {name_feature}. Choose one of the following "
+            msg += "features:\n"
+            msg += self.get_string_all_features()
+            raise InvalidArgument(msg) from err
+        return feature_class()
+
+
+class FeatureTemplate(ABC):
+    """Skeleton for a generic feature."""
+    method_to_query = ""
+
+    @abstractmethod
+    def __init__(self, name):
+        self.name = name
+        self.args = {}
+
+    def query(self, system, **kwargs):
+        debug = kwargs.get("debug", False)
+        self.args.update(system.export_attributes_to_source())
+        self.args.update(kwargs)
+        try:
+            self.source_methods = select_source_method(system,
+                                                       self.method_to_query,
+                                                       **kwargs)
+        except NoSupportedSourcesFound as exception:
+            # if no sources are found in the system for this
+            # particular query, jump to the next one without
+            # stopping execution
+            LOG.error(exception, exc_info=debug)
+            return
```

Comment:
Careful, I do not know if you intend for 'None' to be a possible answer. I would always try to return an empty array, removes the need for ifs down the line.

Reply:
Yes, this was done in purpose. At first I did have some empty collection for such case, but it made it difficult to distinguish between the case where a system could not be queried from the cases when a system was successfully queried but the feature was not found.

Reply:
Ok, if it is on purpose we can close this one.